docs/protocol.md

```diff
@@ -1,26 +1,83 @@
+connecting the two.
+
+Consider those [SOA][soa] problems again:
+
+ - Service discovery: All producers and consumers register themselves with the
+   routing mesh. Consumers access producers only by their name; no need to know
+   about hosts or ports.
+
+ - Fault tolerance: The routing mesh tracks things like failure rates and
+   [SLA][sla] violations. It can intelligently detect unhealthy hosts and
+   remove them from the pool of available hosts.
+
+ - Request tracing: Built into the protocol as a first class citizen.
+
+[sla]: https://en.wikipedia.org/wiki/Service-level_agreement "SLA"
+
+By consolidating logic *between* producers and consumers, this also allows us
+to push core features to our entire SOA without requiring applications to
+update libraries.
+
+Additionally, when developing this protocol and routing mesh, we had the
+following goals in mind:
+
+- The protocol must be easy to implement in multiple languages, especially
+  Javascript, Python, and Go. We do not want to restrict ourselves to one
+  language.
+
+- Async behavior is fundamental. We follow a traditional request / response
+  model with out of order responses; slow requests must not block subsequent,
+  faster requests at the head of the line.
 
-- Easy to implement in multiple languages, especially JS and Python
-- High performance forwarding path. Intermediaries can make a forwarding
-  decision quickly.
-- Request / response model with out of order responses. Slow requests will not
-  block subsequent faster requests at head of line.
 - Large requests/responses may/must be broken into fragments to be sent
-  progressively.
-- Optional checksums
-- Can be used to transport multiple protocols between endpoints, e.g. HTTP+JSON
-  and Thrift.
+  progressively (i.e., "streaming" requests are supported).
+
+- The protocol can be used with multiple transports, e.g. JSON and Thrift,
+  although Thrift is preferred for its [IDL][idl].
 
-## Why not finagle-mux?
+- The routing mesh needs a high-performance forwarding path. Intermediaries
+  must be able to make a forwarding decision quickly.
 
-The Finagle system developed by Twitter was a big inspiration for this
-protocol, specifically finagle-mux. However, there are several additional
-features we want out of this system that will require changes. Also, there is
-only one implementation of finagle-mux, which is in Scala. At Uber we'll need
-to implement at least a node and python version of this protocol, and a Go and
-JVM version are likely to follow after.
+- Optional checksums for request/reponse integrity.
+
+[idl]: https://thrift.apache.org/docs/idl "IDL"
+
+### Why not finagle-mux?
+
+The [Finagle][finagle] system developed by Twitter was a big inspiration for
+this protocol, specifically [finagle-mux][mux]. However, there are several
+additional features we want out of this system that will require changes. Also,
+there is only one implementation of finagle-mux, which is in Scala. At Uber
```

Comment:
I'm in favor of removing this. "Several additional features we want" is too generic. The features we *do* want are addressed above, and the language limitation is also addressed above.

PS: Here, are, some, more, commas, for, you, because, I, know, you, love, them :P